utilities/include/alps/numeric/tensors/mpi/mpi_shared_allocator.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 1998-2018 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+
+#ifndef ALPSCORE_MPI_SHARED_ALLOCATOR_HPP
+#define ALPSCORE_MPI_SHARED_ALLOCATOR_HPP
+
+#include <alps/config.hpp>
+
+#ifdef ALPS_HAVE_MPI
+#include <mpi.h>
+
+#if MPI_VERSION > 2
+#define ALPS_HAVE_SHARED_ALLOCATOR
+#else
+#ifndef NDEBUG
+#warning "MPI shared memory is not supported by your MPI library."
+#endif
+#endif
+
+#ifdef ALPS_HAVE_SHARED_ALLOCATOR
+namespace alps {
+  namespace numerics {
+    namespace detail {
+      template <typename T>
+      class mpi_shared_allocator {
+      public:
+
+        mpi_shared_allocator(MPI_Comm& comm, MPI_Win& win) : comm_(&comm), win_(&win) {
+
+        };
+
+        T* allocate(size_t size) {
+          void *mem;
+          // compute local size for array
+          size_t nloc = local_size(size);
+          // allocate shared memory
+          MPI_Win_allocate_shared(nloc * sizeof(T), sizeof(T), MPI_INFO_NULL, *comm_, &mem, win_);
```

Comment:
Also, here you override the stored window in `win_`, so why do you need the parameter in the constructor?

Reply:
I agree that the shared memory semantic should be revised, since it was just scratch implementation and I never use it in any codes, and I will probably for now remove it from this PR since I am too lazy to think about it right now.

The idea was the following. 
1. Allocation and deallocation should be done through MPI window.  This is the only collective operations.
2. When you access data for writing you should lock window
3. Reading can be done as simple "operator[]" access
4. When you copy tensor you can't copy data, since it's shared-allocated and bound to window, so, you need to copy pointer to the underlying data (this is definitely not implemented), and copy pointer to the window. 
5. When you destroy the tensor, your memory should not be released, since it is still connected to the window. And you still get access through MPI_Get, MPI_Put methods.
6. When the shared memory region is not needed anymore you should free this window (see 1.)

Reply:
That seems very reasonable.  As far as I understand it, a shared memory is always tied to a window, correct?  In this case, how about bundling the two together, IOW, the allocator returns something like this:

```C++
template <typename T> struct mpi_shared_memory {
      MPI_Win window;
      T *memory;
 };
```

So in this case you can make sure that you always destroy the window with it.  Would that work?